pkg/sync/syncer.go

```diff
@@ -1452,6 +1472,532 @@ func (s *syncer) syncGrantsForResource(ctx context.Context, resourceID *v2.Resou
+
+	skipEGForResourceType := make(map[string]bool)
+
+	filterEntitlement, err := s.externalResourceReader.GetEntitlement(ctx, &reader_v2.EntitlementsReaderServiceGetEntitlementRequest{
+		EntitlementId: entitlementId,
+	})
+	if err != nil {
+		return err
+	}
+
+	grants, err := s.listExternalGrantsForEntitlement(ctx, filterEntitlement.GetEntitlement())
+	if err != nil {
+		return err
+	}
+
+	ents := make([]*v2.Entitlement, 0)
+	principals := make([]*v2.Resource, 0)
+	resourceTypes := make([]*v2.ResourceType, 0)
+	resourceTypeIDs := mapset.NewSet[string]()
+	resourceIDs := make(map[string]*v2.ResourceId)
+
+	grantsForEnts := make([]*v2.Grant, 0)
+
+	for _, g := range grants {
+		resourceTypeIDs.Add(g.Principal.Id.ResourceType)
+		resourceIDs[g.Principal.Id.Resource] = g.Principal.Id
+	}
+
+	for _, resourceTypeId := range resourceTypeIDs.ToSlice() {
+		resourceTypeResp, err := s.externalResourceReader.GetResourceType(ctx, &reader_v2.ResourceTypesReaderServiceGetResourceTypeRequest{ResourceTypeId: resourceTypeId})
+		if err != nil {
+			return err
+		}
+		// Should we error or skip if this is not user or group?
+		for _, t := range resourceTypeResp.ResourceType.Traits {
+			if t == v2.ResourceType_TRAIT_USER || t == v2.ResourceType_TRAIT_GROUP {
+				resourceTypes = append(resourceTypes, resourceTypeResp.ResourceType)
+				continue
+			}
+		}
+
+		rtAnnos := annotations.Annotations(resourceTypeResp.ResourceType.Annotations)
+		skipEntitlements := rtAnnos.Contains(&v2.SkipEntitlementsAndGrants{})
+		skipEGForResourceType[resourceTypeResp.ResourceType.Id] = skipEntitlements
+	}
+
+	for _, resourceId := range resourceIDs {
+		resourceResp, err := s.externalResourceReader.GetResource(ctx, &reader_v2.ResourcesReaderServiceGetResourceRequest{ResourceId: resourceId})
+		if err != nil {
+			return err
+		}
+		resourceVal := resourceResp.GetResource()
+		resourceAnnos := annotations.Annotations(resourceVal.GetAnnotations())
+		batonID := &v2.BatonID{}
+		resourceAnnos.Update(batonID)
+		resourceVal.Annotations = resourceAnnos
+		principals = append(principals, resourceVal)
+	}
+
+	for _, principal := range principals {
+		skipEnts := skipEGForResourceType[principal.Id.ResourceType]
+		if skipEnts {
+			continue
+		}
+		resourceEnts, err := s.listExternalEntitlementsForResource(ctx, principal)
+		if err != nil {
+			return err
+		}
+		ents = append(ents, resourceEnts...)
+	}
+
+	for _, ent := range ents {
+		grantsForEnt, err := s.listExternalGrantsForEntitlement(ctx, ent)
+		if err != nil {
+			return err
+		}
+		grantsForEnts = append(grantsForEnts, grantsForEnt...)
+	}
+
+	err = s.store.PutResourceTypes(ctx, resourceTypes...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutResources(ctx, principals...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutEntitlements(ctx, ents...)
+	if err != nil {
+		return err
+	}
+
+	err = s.store.PutGrants(ctx, grantsForEnts...)
+	if err != nil {
+		return err
+	}
+
+	l.Info("Synced external resources",
```

Comment:
It'd be nice to log the external entitlement ID here too.
```suggestion
	l.Info("Synced external resources for entitlement",
	zap.String("external_entitlement_id", entitlementId),
```